README.md
# ltermio - A Lightweight POSIX terminal I/O library

The package contains 5 modules: ***cursor***, ***termkey***, ***termouse***, ***color256*** and ***unicon***. Tested only on **MacOS terminal** and **iTerm2**, so the platform compatibility has not been well verfied yet.  
All functions are based on **XTerm** specification, **CSI** sequences and **termios**, no additional requirements other than the standard library.

**Applicability**: They are only for **POSIX** terminal applications.

References:  
<https://en.wikipedia.org/wiki/ANSI_escape_code>  
<https://www.xfree86.org/current/ctlseqs.html>

## Installation & Usage
Uses pip to install the package:

`pip3 install ltermio`

More details on module usage:

`pydoc ltermio.module-name`

## History & Why
As we all know, character terminal are outdated facilities, why did I still write such a package?

When I was learning **Python** a few months ago, I decided to write a terminal **Tetris** game as a practice of the language learning. Due to the learning reason, I did not want to use any third-party packages. So when I finally finished the game, there naturally formed this by-product.

## color256 module
Sets 256-color display attributes of character terminal.

There are three pairs of functions to set or restore colors according to their literal name:

```python
    set_fcolor(color: int)
    reset_fcolor()
    set_bcolor(color: int)
    reset_bcolor()
    set_color(fcolor: int, bcolor: int)
    reset_color()
```

Enum class *Color* defined constants of some common-used colors and grayscales, and more colors can be generated by function *make_color(r, g, b)*.

A simple sample as following:

```python
    import ltermio
    from ltermio import Color

    ltermio.set_color(Color.GOLD, Color.BLUE)
    print('Hello, ltermio!')
    color = ltermio.make_color(4, 3, 1) # get a bronze color
    ltermio.set_fcolor(color)
    ...
    ltermio.reset_color()
```

## cursor module
Wrapper functions of the **CSI(Control Sequence Introducer)** sequences about cursor and screen.

The sequences of following functions based on are not part of the **ANSI** specification, but be supported by most of terminals.

```python
    show_cursor()
    hide_cursor()
    switch_screen()
    restore_screen()
```

A several of additional functions are provided for text composing:

```python
    v_composing(seq: str) -> str
    downward_seq(text: str, cols: int) -> str
    vert_seq(text: str) -> str
    rect_border_seq(width: int, height: int, sym: str) -> str
```

Following sample code of *v_composing()* displays three big colorful greetings:

```python
    from ltermio import putmsg, v_composing

    _P1 = ':{0}\x1b3hj'
    _P15 = ':{0}\x1b6l:{0}\x1b11hj'
    _DASH = '5:{0}\x1b11hj'
    _NEXT_POS = '5k17l'

    LETTERS = {
        'E': v_composing(f'{(_DASH + _P1) * 2}{_DASH}{_NEXT_POS}'),
        'H': v_composing(f'{_P15 * 2}{_DASH}{_P15 * 2}{_NEXT_POS}'),
        'L': v_composing(f'{_P1 * 4}{_DASH}{_NEXT_POS}'),
        'O': v_composing(f'{_DASH}{_P15 * 3}{_DASH}{_NEXT_POS}'),
    }

    greeting = ''.join(map(LETTERS.get, 'HELLO'))
    putmsg(3, 20, greeting.format('\u2b50'))
    putmsg(9, 14, greeting.format('\u2b55'))
    putmsg(15, 8, greeting.format('\U0001f7e2'))
```

## termkey module
Functions to read input in non-canonical mode.

There are 5 ***curses*** like functions:

+ *getch()*: Gets a character from stdin.
+ *ungetch()*: Puts one or more characters into the key buffer.
+ *getkey()*: Calls getch() and transforms character to keycode.
+ *ungetkey()*: Puts a keycode into the key buffer.
+ *setparams()*: Sets frenquently-used attributes of the input.

And a function to support mouse tracking.

+ *mouse_handler()*: Sets a function to handle mouse report.

Function keycodes in common using are defined in enum class *Key*.

A typical usage example as following:

```python
    import ltermio
    from ltermio import Key

    ltermio.setparams(echo=False)
    key = ltermio.getkey()
    while key != Key.ESC:
        ...
        key = ltermio.getkey()
    ltermio.setparams()
```

## termouse module
Detects and reports mouse events.

The implementation of the module follows the **XTerm** specification of the mouse tracking and uses normal tracking mode, see also:

<https://www.xfree86.org/current/ctlseqs.html#Mouse%20Tracking>
    
Calls *ltermio.mouse_tracking_on()* to turn on the mouse tracking, or sets the parameter `mouse=True` on the decorator *ltermio.appentry_args()*.

The mouse events are reported by *ltermio.getkey()*, with value encoded in a 32-bits integer and larger than *Key.MOUSE_EVENT*. It is simple to make difference from normal key codes by `code > Key.MOUSE_EVENT`.  
Decodes the event codes by calling *ltermio.decode_mouse_event()* which returns a tuple with explicit items.
        
An example to get key and mouse inputs as following:

```python
    from ltermio import Key, MouseEvent

    ltermio.mouse_tracking_on()
    ltermio.setparams(echo=False)

    code = ltermio.getkey()
    while code != Key.CONTROL_X:
        if code > Key.MOUSE_EVENT:
            event, row, col, modifiers = ltermio.decode_mouse_event(code)
            if event == MouseEvent.B1_CLICKED:
                ... # do something with mouse event
        else:
            ... # do something with key input
        code = ltermio.getkey()

    ltermio.setparams(echo=True)
    ltermio.mouse_tracking_off()
```

## unicon module
Collection of some common icons in unicode character set.


CHANGELOG
0.4.10 Current release
  * Change the modifier values that defined in Key, to prevent conflicts.
  * Prefix 'r' for all docstrings of the package to assure WYSWYG.
  + Add more docstrings for mouse module.
  + Add module docstrings to README.md.
0.4.8
  * Optimize the implement of termouse module.
  * Modified the import pattern in __init__.py.
  * Change some descriptions of comments.
0.4.7
  + Add doc comments for module termouse.py
0.4.6
  * Fix appentry can't set parameters.
  * Adjust the order of (row, col) in decode_mouse_event() return value.
0.4.5
  + Roughly implemented mouse supporting.
0.4.0
  + Add supporting for mouse event report in getkey().
  + Add a function mouse_handler() to set mouse report handler.
  + Add a file termouse.py, get ready to add mouse supporting.
0.3.5
  + Add ungetch() in termkey module.
  + Add ungetkey() in termkey module.
  * Lightly modify the implementation of the getkey().
0.3.2
  + Add comment for decorator appentry().
  + Add reset_color() after return from entry function in appentry().
0.3.1
  * fix bug about appentry in __init__.py.
0.3.0
  * Change build backend from 'setuptools' to 'hatchling'.
  * Rename (re)set_foreground_color to shorter name (re)set_fcolor.
  * Rename (re)set_background_color to shorter name (re)set_bcolor.
  + Add a decorator appentry() in __init__.py.
  + Import all modules' public functions and classes in __init__.py.
0.2.7
  * Rename keyparam() to setparams().
  * Modify relative comments and README.md.
0.2.6
  - Remove the features of disable echo and disable intr from getch().
  + Add a new function keyparam() to set echo and intr attribute.
  * Modify relative comments.
0.1.6
  * Change the building script from setup.py to pyproject.toml.
0.1.5
  * Update some metadata in setup.py.
  * Update README.md.
  + Add license info to source code.
0.1.4
  * Optimize a few of code and adjust some layout format.
0.1.3
  * Fix the getch() function has not recovered the setting of the stdin.
0.1.2
  + The first version with complete functionality.


src/cursor.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

#   Copyright 2023 Brooks Su
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.

"""Wrapper functions of CSI(Control Sequence Introducer) sequences about
cursor and screen.

The sequences of following functions based on are not part of the ANSI
specification, but be supported by most of terminals.

    show_cursor()
    hide_cursor()
    switch_screen()
    restore_screen()

A several of additional functions are provided for text composing:

    v_composing()
    downward_seq()
    vert_seq()
    rect_border_seq()

Reference: https://en.wikipedia.org/wiki/ANSI_escape_code
           https://www.xfree86.org/current/ctlseqs.html
"""

from wcwidth import wcswidth


_CSI_LEAD = '\033['


def cursor_up_seq(num: int = 1):
    """Returns a sequence string that can move the cursor up 'num' rows.
    """
    return f'{_CSI_LEAD}{num}A'


def cursor_up(num: int = 1):
    """Moves the cursor up 'num' rows.
    """
    print(cursor_up_seq(num), end='', flush=True)


def cursor_down_seq(num: int = 1):
    """Returns a sequence string that can move the cursor down 'num' rows.
    """
    return f'{_CSI_LEAD}{num}B'


def cursor_down(num: int = 1):
    """Moves the cursor down 'num' rows.
    """
    print(cursor_down_seq(num), end='', flush=True)


def cursor_right_seq(num: int = 1):
    """Returns a sequence string that can move the cursor 'num' columns
    to right.
    """
    return f'{_CSI_LEAD}{num}C'


def cursor_right(num: int = 1):
    """Moves the cursor 'num' columns to right.
    """
    print(cursor_right_seq(num), end='', flush=True)


def cursor_left_seq(num: int = 1):
    """Returns a sequence string that can move the cursor 'num' columns
    to left.
    """
    return f'{_CSI_LEAD}{num}D'


def cursor_left(num: int = 1):
    """Moves the cursor 'num' columns to left.
    """
    print(cursor_left_seq(num), end='', flush=True)


def cursor_pos_seq(row: int, col: int):
    """Returns a sequence string that can move the cursor to the coordinate
    (row, col).
    """
    return f'{_CSI_LEAD}{row};{col}H'


def cursor_pos(row: int, col: int):
    """Moves the cursor to the coordinate (row, col).
    """
    print(cursor_pos_seq(row, col), end='', flush=True)


def putmsg(row: int, col: int, message: str, *args, **kwargs):
    """Moves the cursor to the coordinate (row, col), and prints message.
    """
    print(f'{cursor_pos_seq(row, col)}{message}',
          *args, **kwargs, end='', flush=True)


def clear_screen():
    """Clears current screen buffer.
    """
    print(f'{_CSI_LEAD}2J', end='', flush=True)


def scroll_up(num: int = 1):
    """Scroll whole page up by 'num' rows. New rows are added at bottom.
    """
    print(f'{_CSI_LEAD}{num}S', end='', flush=True)


def scroll_down(num: int = 1):
    """Scroll whole page down by 'num' rows. New rows are added at top.
    """
    print(f'{_CSI_LEAD}{num}T', end='', flush=True)


def show_cursor():
    """Shows the cursor on screen.
    """
    print(f'{_CSI_LEAD}?25h', end='', flush=True)


def hide_cursor():
    """Hides the cursor from screen.
    """
    print(f'{_CSI_LEAD}?25l', end='', flush = True)


def switch_screen():
    """Uses alternate screen buffer to display message, it makes the content
    in the normal screen buffer remaining unchange.
    """
    print(f'{_CSI_LEAD}?47h', end='', flush=True)


def restore_screen():
    """Switches back to the normal screen buffer.
    """
    print(f'{_CSI_LEAD}?47l', end='', flush=True)


_comp_funcs = {
    'h': cursor_left_seq,
    'l': cursor_right_seq,
    'k': cursor_up_seq,
    'j': cursor_down_seq,
}

def v_composing(comp_seq: str) -> str:
    r"""Generates text composing sequence.

    The parameter 'comp_seq' can use a series of VI like commands(hlkj)
    to specify relative cursor movements, using ':' command to insert
    text, and an ESC(\x1b) to back command mode. A numberic prefix can be
    added to the command to specify repetition times.

    Following sample code displays three big colorful greetings:

        from cursor import putmsg, v_composing

        _P1 = ':{0}\x1b3hj'
        _P15 = ':{0}\x1b6l:{0}\x1b11hj'
        _DASH = '5:{0}\x1b11hj'
        _NEXT_POS = '5k17l'

        LETTERS = {
            'E': v_composing(f'{(_DASH + _P1) * 2}{_DASH}{_NEXT_POS}'),
            'H': v_composing(f'{_P15 * 2}{_DASH}{_P15 * 2}{_NEXT_POS}'),
            'L': v_composing(f'{_P1 * 4}{_DASH}{_NEXT_POS}'),
            'O': v_composing(f'{_DASH}{_P15 * 3}{_DASH}{_NEXT_POS}'),
        }

        greeting = ''.join(map(LETTERS.get, 'HELLO'))
        putmsg(3, 20, greeting.format('\u2b50'))
        putmsg(9, 14, greeting.format('\u2b55'))
        putmsg(15, 8, greeting.format('\U0001f7e2'))

    Returns:
        A text composing sequnce that can be output to the screen in the
        expected shape.

    Raises:
        ValueError: Invalid composing sequence.
    """
    global _comp_funcs
    result = []
    index = repeat = 0

    while index < len(comp_seq):
        cmd = comp_seq[index]
        index += 1
        if cmd.isdigit():
            repeat = repeat * 10 + int(cmd)
        elif cmd in _comp_funcs:
            result.append(_comp_funcs[cmd](repeat if repeat else 1))
            repeat = 0
        elif cmd == ':':  # In insertion mode
            esc = comp_seq.find('\x1b', index)
            if esc < 0:
                result.append(comp_seq[index:] * (repeat if repeat else 1))
                break
            result.append(comp_seq[index:esc] * (repeat if repeat else 1))
            repeat = 0
            index = esc + 1
        elif cmd != '\x1b':
            raise ValueError('Invalid composing sequence')
        # else: ignores redundant ESC(\x1b)
    return ''.join(result)


def downward_seq(text: str, back_cols: int) -> str:
    r"""Generates a text sequence that makes subsequent text to be output
    in downward direction.

    After the result text be output to screen, the cursor takes 'back_cols'
    columns back from the end position of the output and moves to the next
    line.

    The following example code draws a triangle on screen:

        putmsg(1, 40, downward_seq('\u2b50', 3) * 24)
        putmsg(1, 40, downward_seq('\u2b50', 1) * 24)
        putmsg(24, 17, '\u2b50' * 24)
    """
    return v_composing(f':{text}\x1b{back_cols}hj')


def vert_seq(text:str) -> str:
    """Generates a text sequence that makes subsequent text be output in
    vertical downward direction.

    It is just a simple wapper of the downward_seq().
    """
    return downward_seq(text, wcswidth(text))


def rect_border_seq(width: int, height: int, sym: str) -> str:
    """Generates a text sequence that draws a rectangle border.
    """
    sym_w = wcswidth(sym)
    hori_border = f':{sym * width}\x1b{sym_w * width}hj'
    vert_borders = ((f':{sym}\x1b{sym_w * (width - 2)}l'
                     f':{sym}\x1b{sym_w * width}hj') * (height - 2))
    return v_composing(hori_border + vert_borders + hori_border)


def _test_cursor():
    def greeting_show():
        p10 = ':{0}\x1b3hj'
        p00001 = '8l:{0}\x1b11hj'
        p10001 = ':{0}\x1b6l:{0}\x1b11hj'
        p1001 = ':{0}\x1b4l:{0}\x1b9hj'
        dash = '5:{0}\x1b11hj'
        next_pos = '5k17l'
        letters = {
            'C': v_composing(f'{dash}{p10 * 3}{dash}{next_pos}'),
            'E': v_composing(f'{(dash + p10) * 2}{dash}{next_pos}'),
            'H': v_composing(f'{p10001 * 2}{dash}{p10001 * 2}{next_pos}'),
            'L': v_composing(f'{p10 * 4}{dash}{next_pos}'),
            'O': v_composing(f'{dash}{p10001 * 3}{dash}{next_pos}'),
            'R': v_composing(f'{dash}{p10001}{dash}{p1001}{p10001}{next_pos}'),
            'S': v_composing(f'{dash}{p10}{dash}{p00001}{dash}{next_pos}'),
            'U': v_composing(f'{p10001 * 4}{dash}{next_pos}'),
        }
        hello = ''.join(map(letters.get, 'HELLO'))
        cursor = ''.join(map(letters.get, 'CURSOR'))
        putmsg(4, 20, hello.format('\U0001f7e4'))
        putmsg(10, 14, hello.format('\U0001f7e0'))
        putmsg(16, 8, cursor.format('\U0001f7e1'))

    def frame_show():
        putmsg(4, 2, rect_border_seq(39, 18, '\u2b50'))
        putmsg(1, 40, downward_seq('\U0001f339', 3) * 24)
        putmsg(1, 40, downward_seq('\U0001f339', 1) * 24)
        putmsg(24, 17, '\U0001f339' * 24)

    switch_screen()
    clear_screen()
    hide_cursor()
    try:
        greeting_show()
        cursor_pos(24, 1)
        input('Enter to continue.')
        clear_screen()
        frame_show()
        cursor_pos(24, 1)
        input('Enter to exit.')
    finally:
        show_cursor()
        restore_screen()


if __name__ == '__main__':
    _test_cursor()


ltermio/__init__.py
# -*- coding: utf-8 -*-

#   Copyright 2023 Brooks Su
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.

r"""A lightweight POSIX character terminal I/O library.

Repo - https://github.com/brookssu/ltermio.git
"""

import functools

from .color256 import (
    make_color,
    Color,
    set_fcolor,
    reset_fcolor,
    set_bcolor,
    reset_bcolor,
    set_color,
    reset_color,
)
from .cursor import (
    cursor_up,
    cursor_down,
    cursor_right,
    cursor_left,
    cursor_pos,
    putmsg,
    clear_screen,
    scroll_up,
    scroll_down,
    show_cursor,
    hide_cursor,
    switch_screen,
    restore_screen,
    v_composing,
    downward_seq,
    vert_seq,
    rect_border_seq,
)
from .termkey import (
    getch,
    ungetch,
    Key,
    ungetkey,
    setparams,
    getkey,
)
from .termouse import (
    mouse_tracking_on,
    mouse_tracking_off,
    MouseEvent,
    set_mouse_mask,
    set_click_interval,
    decode_mouse_event,
)
from .unicon import UnicodeIcon as UIcon


__version__ = '0.4.10'
__all__ = ['cursor', 'termkey', 'color256', 'unicon', 'termouse']


def appentry(*, echo=False, intr=False, cursor=False, mouse=False):
    r"""A decorator of the ltermio application entry.

    Before enters entry function, the decorator switchs and clears screen,
    sets cursor and other input attributes according to the parameters.
    And while the entry function returns, the decorator restores screen,
    cursor, color and other set attributes.

    Args:
        echo: Echoes input characters if True, False to disable.
        intr: False to disable keyboard interrupt signals.
        cursor: False to hide cursor while True to show.
        mouse: True to enable mouse tracking.
    """
    def wrapper(func):
        @functools.wraps(func)
        def inner(*args, **kwargs):
            switch_screen()
            clear_screen()
            setparams(echo=echo, intr=intr)
            if not cursor:
                hide_cursor()
            if mouse:
                mouse_tracking_on()
            try:
                return func(*args, **kwargs)
            finally:
                mouse_tracking_off()
                show_cursor()
                reset_color()
                setparams()
                restore_screen()
        return inner
    return wrapper


ltermio/termouse.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

#   Copyright 2023 Brooks Su
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.

r"""Detects and reports mouse events.

The implementation of the module follows the XTerm specification of the
mouse tracking and uses normal tracking mode, see also:

    https://www.xfree86.org/current/ctlseqs.html#Mouse%20Tracking

Calls ltermio.mouse_tracking_on() to turn on the mouse tracking, or sets
the parameter `mouse=True` on the decorator ltermio.appentry_args().

The mouse events are reported by ltermio.getkey(), with value encoded in
a 32-bits integer and larger than Key.MOUSE_EVENT. It is simple to make
difference from normal key codes by `code > Key.MOUSE_EVENT`.
Decodes the event codes by calling ltermio.decode_mouse_event() which
returns a tuple with explicit items.

An example to get key and mouse inputs as following:

    from ltermio import Key, MouseEvent

    ltermio.mouse_tracking_on()
    ltermio.setparams(echo=False)

    code = ltermio.getkey()
    while code != Key.CONTROL_X:
        if code > Key.MOUSE_EVENT:
            event, row, col, modifiers = ltermio.decode_mouse_event(code)
            if event == MouseEvent.B1_CLICKED:
                ... # do something with mouse event
        else:
            ... # do something with key input
        code = ltermio.getkey()

    ltermio.setparams(echo=True)
    ltermio.mouse_tracking_off()
"""

import time
from enum import IntEnum

from .termkey import getkey, setparams, mouse_handler, Key


# Constants of mouse tracking modes
_NORMAL_TRACKING = 1000
#_HILITE_TRACKING = 1001
#_CELL_MOTION_TRACKING = 1002
#_ANY_MOTION_TRACKING = 1003


class MouseEvent(IntEnum):
    r"""Enum constants of the mouse events.
    """
    B1_PRESSED = 0x0002_0000
    B1_RELEASED = 0x0004_0000
    B1_CLICKED = 0x0008_0000
    B2_PRESSED = 0x0010_0000
    B2_RELEASED = 0x0020_0000
    B2_CLICKED = 0x0040_0000
    B3_PRESSED = 0x0080_0000
    B3_RELEASED = 0x0100_0000
    B3_CLICKED = 0x0200_0000
    B4_PRESSED = 0x0400_0000
    B4_RELEASED = 0x0800_0000
    B4_CLICKED = 0x1000_0000
    B5_PRESSED = 0x2000_0000
    B5_RELEASED = 0x4000_0000
    B5_CLICKED = 0x8000_0000


_MOUSE_EVENTS = (
    (MouseEvent.B1_PRESSED, MouseEvent.B1_RELEASED, MouseEvent.B1_CLICKED),
    (MouseEvent.B2_PRESSED, MouseEvent.B2_RELEASED, MouseEvent.B2_CLICKED),
    (MouseEvent.B3_PRESSED, MouseEvent.B3_RELEASED, MouseEvent.B3_CLICKED),
    (MouseEvent.B4_PRESSED, MouseEvent.B4_RELEASED, MouseEvent.B4_CLICKED),
    (MouseEvent.B5_PRESSED, MouseEvent.B5_RELEASED, MouseEvent.B5_CLICKED),
)
_PRESSED_INDEX = 0
_RELEASED_INDEX = 1
_CLICKED_INDEX = 2


# pylint: disable=invalid-name
# pylint: disable=global-statement
_mouse_mask: int = 0xfffe_0000
_pressed_button: int = 0
_pressed_time: float = 0.
_click_interval: float = 0.2  # seconds


def _tracking_on(mode: int):
    # DEC private mode set(DECSET)
    print(f'\033[?{mode}h', end='', flush=True)


def _tracking_off(mode: int):
    # DEC private mode reset(DECRST)
    print(f'\033[?{mode}l', end='', flush=True)


def mouse_tracking_on():
    r"""Truns on mouse tracking.
    """
    mouse_handler(_on_mouse_event)
    _tracking_on(_NORMAL_TRACKING)


def mouse_tracking_off():
    r"""Truns off mouse tracking.
    """
    _tracking_off(_NORMAL_TRACKING)
    mouse_handler(None)


def set_mouse_mask(mask: int):
    r"""Sets the mouse events to be reported.

    If turns mouse tracking on and does not call this function, all
    mouse events are reported by default.

    Args:
        mask: A bitwise combination of the mouse events, e.g.
            MouseEvent.B1_CLICKED + MouseEvent.B3_CLICKED
    """
    global _mouse_mask
    _mouse_mask = mask


def set_click_interval(interval: float):
    r"""Sets the maximum time limit between press and release in order
    that code can recognize them as one click.

    When a CLICKED event occurs, it also means that a RELEASED event
    has occurred at the same time.  In this case, ltermio does not
    report two events but only reports the CLICKED, or only reports
    the RELEASED when the CLICKED event is masked.

    Args:
        interval: time interval in seconds, the default value is 0.2s.
    """
    global _click_interval
    _click_interval = interval


def decode_mouse_event(code: int) -> tuple[int, int, int, int]:
    r"""Decodes the mouse event code into a tuple.

    Args:
        code: The mouse event code that returns by ltermio.getkey(), which
            can be determined by `code > Key.MOUSE_EVENT`.

    returns:
        A tuple with 4 items: (event, row, col, modifiers):
        event: One of the mouse events which defined in the MouseEvent.
        row, col: Screen coordinate of the mouse when the event occurs.
        modifiers: Modifier keys(Shift, Alt or Meta, Control) that be
            pressed when the event occurs, their values are identical
            to their values in ltermio.Key.
    """
    return (code & 0xfffe0000,  # event
            (code & 0x3f80) >> 7,  # row
            code & 0x7f,  # col
            (code & 0x1c000) >> 2)  # modifiers


def _on_mouse_event(data: int, col: int, row: int) -> int:
    global _pressed_button, _pressed_time

    # On button press or release, xterm sends CSI M Cb Cx Cy. The low two
    # bits of Cb(data) encode button information:
    #       0=B1 pressed, 1=B2 pressed, 2=B3 pressed, 3=release.
    # The next three bits encode the modifiers which were down when the
    # button was pressed and are added together:
    #       4=Shift, 8=Alt(Meta), 16=Control.

    button = data & 0x03
    # Encodes modifiers, row and col into lower 17 bits of the event code.
    lower = ((data & 0x1c) << 12) | (row << 7) | col

    if button == 3:  # Button released
        if time.perf_counter() - _pressed_time < _click_interval:
            event = _MOUSE_EVENTS[_pressed_button][_CLICKED_INDEX]
            if _mouse_mask & event:
                return event | lower
        event = _MOUSE_EVENTS[_pressed_button][_RELEASED_INDEX]
        return (event | lower) if (_mouse_mask & event) else Key.NONE

    # Button pressed
    if data >= 64:
        # Wheel mice may return buttons 4 and 5. Those buttons are represented
        # by the same event codes as buttons 1 and 2 respectively, except that
        # 64 is added to the event code. Release events for the wheel buttons
        # are not reported.
        if button > 1:
            return Key.NONE
        button += 3
    _pressed_button = button
    _pressed_time = time.perf_counter()
    event = _MOUSE_EVENTS[button][_PRESSED_INDEX]
    return (event | lower) if (_mouse_mask & event) else Key.NONE


def _mdf_names(code):
    modifiers = []
    if code & Key.SHIFT:
        modifiers.append(Key.SHIFT.name)
    if code & Key.ALT:
        modifiers.append(Key.ALT.name)
    if code & Key.CONTROL:
        modifiers.append(Key.CONTROL.name)
    if code & Key.META:
        modifiers.append(Key.META.name)
    return ' '.join(modifiers)


def _test_termouse():
    mouse_tracking_on()
    setparams(echo=False, intr=False)
    print('Press any key or click mouse to get code, CONTROL-X to exit.')
    code = getkey()
    while code != Key.CONTROL_X:
        if code < Key.MOUSE_EVENT:
            print(f'Key code({code:04X}) - char({chr(code)!r})'
                  f' {_mdf_names(code)}')
        else:
            event, row, col, modifiers = decode_mouse_event(code)
            print(f'Mouse code({code:08X}) - {MouseEvent(event).name}'
                  f'({row}, {col}) {_mdf_names(modifiers)}')
        code = getkey()
    setparams()
    mouse_tracking_off()


if __name__ == '__main__':
    _test_termouse()


ltermio/color256.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

#   Copyright 2023 Brooks Su
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.

r"""Sets 256-color display attributes of character terminal.

There are three pairs of functions to set or restore colors according
to their literal name:

    set_fcolor(color)
    reset_fcolor()

    set_bcolor(color)
    reset_bcolor()

    set_color()
    reset_color()

Enum class Color defined constants of some basic colors and grayscales,
and more colors can be generated by function make_color(r, g, b).

A simple sample as following:

    import ltermio
    from ltermio import Color

    ltermio.set_color(Color.GOLD, Color.BLUE)
    print('Hello, ltermio!')
    color = ltermio.make_color(4, 3, 1) # get a bronze color
    ltermio.set_fcolor(color)
    ...
    ltermio.reset_color()

Reference: https://en.wikipedia.org/wiki/ANSI_escape_code
"""

import enum

_CSI_LEAD = '\033['

def make_color(red: int, green: int, blue: int) -> int:
    r"""Returns a 256-color number according to the given RGB parameters.

    Args:
        All parameters' valid values must be in range of 0~5.

    Raises:
        ValueError: Argument value out of range.
    """
    if not (0 <= red <= 5 and 0 <= green <= 5 and 0 <= blue <= 5):
        raise ValueError('Argument value out of range.')
    return 16 + red * 36 + green * 6 + blue


class Color(enum.IntEnum):
    r"""Enumerates some common constants of colors.
    """
    BLACK = 0
    RED = 1
    GREEN = 2
    YELLOW = 3
    BLUE = 4
    MAGENTA = 5
    CYAN = 6
    WHITE = 7
    BRIGHT_BLACK = 8
    BRIGHT_RED = 9
    BRIGHT_GREEN = 10
    BRIGHT_YELLOW = 11
    BRIGHT_BLUE = 12
    BRIGHT_MAGENTA = 13
    BRIGHT_CYAN = 14
    BRIGHT_WHITE = 15

    GRAYSCALE_DARKEST = 232
    GRAYSCALE_DARK = 238
    GRAYSCALE_MIDDLE = 243
    GRAYSCALE_LIGHT = 249
    GRAYSCALE_LIGHTEST = 255
    GRAY = make_color(3, 3, 3)
    SILVER = make_color(4, 4, 4)

    TAN = make_color(5, 4, 3)
    BRONZE = make_color(4, 3, 1)
    COPPER = BRONZE
    COFFEE = make_color(2, 1, 0)
    CHOCOLATE = make_color(4, 2, 1)
    BROWN = make_color(3, 1, 1)

    PINK = make_color(5, 3, 3)
    HOT_PINK = make_color(5, 2, 4)
    DEEP_PINK = make_color(5, 0, 3)
    TOMATO = make_color(5, 2, 2)
    INDIAN_RED = TOMATO
    MAROON = RED
    CRIMSON = make_color(5, 0, 1)
    SCARLET = CRIMSON

    VIOLET = make_color(3, 0, 5)
    BRIGHT_VIOLET = make_color(5, 3, 5)
    ORCHID = make_color(3, 1, 5)
    BRIGHT_ORCHID = make_color(4, 2, 4)
    PLUM = make_color(5, 4, 5)
    LAVENDER = make_color(4, 4, 5)
    MEDIUM_PURPLE = make_color(3, 2, 5)
    PURPLE = make_color(3, 0, 3)

    INDIGO = make_color(1, 0, 4)
    NAVY = make_color(0, 0, 3)
    ROYAL_BLUE = make_color(2, 3, 5)
    SKY_BLUE = make_color(3, 5, 5)
    DEEP_SKY_BLUE = make_color(0, 4, 5)
    AZURE = make_color(4, 5, 5)

    TURQUOISE = make_color(2, 5, 4)
    SPRING_GREEN = make_color(2, 4, 3)
    SEA_GREEN = make_color(1, 4, 4)
    DEEP_SEA_GREEN = make_color(1, 3, 2)
    OLIVE = make_color(3, 3, 0)

    BEIGE = make_color(5, 5, 4)
    IVORY = BEIGE
    KHAKI = make_color(5, 5, 3)
    DEEP_KHAKI = make_color(4, 4, 2)
    GOLD = make_color(5, 4, 0)
    ORANGE = make_color(5, 3, 0)
    DEEP_ORANGE = make_color(5, 2, 0)


def set_fcolor(color: int):
    r"""Sets foreground color.
    """
    print(f'{_CSI_LEAD}38;5;{color}m', end='')


def reset_fcolor():
    r"""Restores the foreground color to the default value.
    """
    print(f'{_CSI_LEAD}39m', end='')


def set_bcolor(color: int):
    r"""Sets background color.
    """
    print(f'{_CSI_LEAD}48;5;{color}m', end='')


def reset_bcolor():
    r"""Restores the background color to the default value.
    """
    print(f'{_CSI_LEAD}49m', end='')


def set_color(fcolor: int, bcolor: int):
    r"""Sets foreground to 'fcolor' and background to 'bcolor'.
    """
    print(f'{_CSI_LEAD}38;5;{fcolor}m{_CSI_LEAD}48;5;{bcolor}m',
          end='')


def reset_color():
    r"""Restores foreground and background color to the default value.
    """
    print(f'{_CSI_LEAD}39;49m', end='')


def _test_color256():
    def color_block(color, width):
        set_bcolor(color)
        print(' ' * width, end='')
        reset_bcolor()

    def text_color():
        for i in range(6):
            for char in 'Hello, make_color!':
                set_fcolor(ord(char) % 32 + 20 + i * 36)
                print(char, end='')
            print('')
        reset_fcolor()

    def grayscale():
        for color in range(Color.GRAYSCALE_DARKEST,
                           Color.GRAYSCALE_LIGHTEST + 1):
            color_block(color, 3)
        print('')

    def rgb():
        for red in range(6):
            for green in range(6):
                for blue in range(6):
                    color_block(make_color(red, green, blue), 2)
            print('')

    def constant_color():
        for color in sorted(Color):
            print(f'{color.name:18s}: {color.value:3d} ', end='')
            color_block(color, 12)
            print('')

    text_color()
    grayscale()
    rgb()
    constant_color()


if __name__ == '__main__':
    _test_color256()
